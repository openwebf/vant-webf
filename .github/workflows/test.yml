name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - dev

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: npm test

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v1
      with:
        token: ${{ secrets.CODECOV_TOKEN }}

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
