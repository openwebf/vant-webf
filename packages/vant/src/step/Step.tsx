import { computed, defineComponent } from 'vue';

// Utils
import { BORDER, createNamespace } from '../utils';
import { STEPS_KEY } from '../steps/Steps';

// Composables
import { useParent } from '@vant/use';

// Components
import { Icon } from '../icon';

const [name, bem] = createNamespace('step');

export default defineComponent({
  name,

  setup(props, { slots }) {
    const { parent, index } = useParent(STEPS_KEY);

    if (!parent) {
      if (process.env.NODE_ENV !== 'production') {
        console.error('[Vant] <Step> must be a child component of <Steps>.');
      }
      return;
    }

    const parentProps = parent.props;

    const getStatus = () => {
      const active = +parentProps.active;
      if (index.value < active) {
        return 'finish';
      }
      return index.value === active ? 'process' : 'waiting';
    };

    const isActive = () => getStatus() === 'process';

    const lineStyle = computed(() => ({
      background:
        getStatus() === 'finish'
          ? parentProps.activeColor
          : parentProps.inactiveColor,
    }));

    const titleStyle = computed(() => {
      if (isActive()) {
        return { color: parentProps.activeColor };
      }
      if (getStatus() === 'waiting') {
        return { color: parentProps.inactiveColor };
      }
    });

    const onClickStep = () => parent.onClickStep(index.value);

    const renderCircle = () => {
      const { iconPrefix, finishIcon, activeIcon, activeColor, inactiveIcon } =
        parentProps;

      if (isActive()) {
        if (slots['active-icon']) {
          return slots['active-icon']();
        }

        return (
          <Icon
            class={bem('icon', 'active')}
            name={activeIcon}
            color={activeColor}
            classPrefix={iconPrefix}
          />
        );
      }

      if (getStatus() === 'finish' && (finishIcon || slots['finish-icon'])) {
        if (slots['finish-icon']) {
          return slots['finish-icon']();
        }

        return (
          <Icon
            class={bem('icon', 'finish')}
            name={finishIcon}
            color={activeColor}
            classPrefix={iconPrefix}
          />
        );
      }

      if (slots['inactive-icon']) {
        return slots['inactive-icon']();
      }

      if (inactiveIcon) {
        return (
          <Icon
            class={bem('icon')}
            name={inactiveIcon}
            classPrefix={iconPrefix}
          />
        );
      }

      return <i class={bem('circle')} style={lineStyle.value} />;
    };

    return () => {
      const status = getStatus();

      return (
        <div
          class={[BORDER, bem([parentProps.direction, { [status]: status }])]}
        >
          <div
            class={bem('title', { active: isActive() })}
            style={titleStyle.value}
            onClick={onClickStep}
          >
            {slots.default?.()}
          </div>
          <div class={bem('circle-container')} onClick={onClickStep}>
            {renderCircle()}
          </div>
          <div class={bem('line')} style={lineStyle.value} />
        </div>
      );
    };
  },
});
