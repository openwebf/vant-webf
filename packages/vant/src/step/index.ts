import { withInstall } from '../utils';
import _Step from './Step';

export const Step = withInstall(_Step);
export default Step;

declare module 'vue' {
  export interface GlobalComponents {
    VanStep: typeof Step;
  }
}
