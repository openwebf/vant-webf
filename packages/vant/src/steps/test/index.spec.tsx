import { mount } from '../../../test';
import { Step } from '../../step';
import { Steps } from '..';

test('should render icon slot correctly', () => {
  const wrapper = mount({
    render() {
      return (
        <Steps active={0}>
          <Step v-slots={{ 'active-icon': () => `Custom Active Icon` }}>B</Step>
          <Step v-slots={{ 'inactive-icon': () => `Custom Inactive Icon` }}>
            A
          </Step>
        </Steps>
      );
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('should emit click-step event when step is clicked', () => {
  const onClickStep = jest.fn();
  const wrapper = mount({
    setup() {
      return () => (
        <Steps active={1} onClick-step={onClickStep}>
          <Step>A</Step>
          <Step>B</Step>
          <Step>C</Step>
        </Steps>
      );
    },
  });

  wrapper.find('.van-step').trigger('click');
  expect(onClickStep).toHaveBeenCalledTimes(0);

  wrapper.find('.van-step__title').trigger('click');
  expect(onClickStep).toHaveBeenCalledWith(0);

  wrapper.findAll('.van-step__circle-container')[2].trigger('click');
  expect(onClickStep).toHaveBeenCalledTimes(2);
  expect(onClickStep).toHaveBeenLastCalledWith(2);
});

test('should change inactive color when using inactive-color prop', () => {
  const wrapper = mount({
    render() {
      return (
        <Steps active={0} inactiveColor="red">
          <Step>A</Step>
          <Step>B</Step>
        </Steps>
      );
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('should change inactive icon when using inactive-icon prop', () => {
  const wrapper = mount({
    render() {
      return (
        <Steps active={0} inactiveIcon="foo">
          <Step>A</Step>
          <Step>B</Step>
        </Steps>
      );
    },
  });

  const steps = wrapper.findAll('.van-step');
  expect(steps[1].find('.van-icon-foo').exists()).toBeTruthy();
  expect(steps[1].html()).toMatchSnapshot();
});

test('should change finish icon when using finish-icon prop', () => {
  const wrapper = mount({
    render() {
      return (
        <Steps active={1} finishIcon="foo">
          <Step>A</Step>
          <Step>B</Step>
        </Steps>
      );
    },
  });

  const firstStep = wrapper.find('.van-step');
  expect(firstStep.find('.van-icon-foo').exists()).toBeTruthy();
  expect(firstStep.html()).toMatchSnapshot();
});

test('should render finish icon slot correctly', () => {
  const wrapper = mount({
    render() {
      return (
        <Steps active={1}>
          <Step v-slots={{ 'finish-icon': () => `Custom Finish Icon` }}>A</Step>
          <Step>B</Step>
        </Steps>
      );
    },
  });

  const firstStep = wrapper.find('.van-step');
  expect(firstStep.html()).toMatchSnapshot();
});

test('should render icon-prefix correctly', () => {
  const wrapper = mount({
    render() {
      return (
        <Steps active={1} iconPrefix="custom-icon">
          <Step>A</Step>
          <Step>B</Step>
        </Steps>
      );
    },
  });

  const steps = wrapper.findAll('.van-step');
  expect(steps[1].html()).toMatchSnapshot();
});
