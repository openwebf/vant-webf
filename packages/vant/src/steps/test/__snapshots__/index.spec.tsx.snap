// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change finish icon when using finish-icon prop 1`] = `
<div class="van-hairline van-step van-step--horizontal van-step--finish">
  <div class="van-step__title">
    A
  </div>
  <div class="van-step__circle-container">
    <i class="van-badge__wrapper van-icon van-icon-foo van-step__icon van-step__icon--finish">
    </i>
  </div>
  <div class="van-step__line">
  </div>
</div>
`;

exports[`should change inactive color when using inactive-color prop 1`] = `
<div class="van-steps van-steps--horizontal">
  <div class="van-steps__items">
    <div class="van-hairline van-step van-step--horizontal van-step--process">
      <div class="van-step__title van-step__title--active">
        A
      </div>
      <div class="van-step__circle-container">
        <i class="van-badge__wrapper van-icon van-icon-checked van-step__icon van-step__icon--active">
        </i>
      </div>
      <div class="van-step__line"
           style="background: red;"
      >
      </div>
    </div>
    <div class="van-hairline van-step van-step--horizontal van-step--waiting">
      <div class="van-step__title"
           style="color: red;"
      >
        B
      </div>
      <div class="van-step__circle-container">
        <i class="van-step__circle"
           style="background: red;"
        >
        </i>
      </div>
      <div class="van-step__line"
           style="background: red;"
      >
      </div>
    </div>
  </div>
</div>
`;

exports[`should change inactive icon when using inactive-icon prop 1`] = `
<div class="van-hairline van-step van-step--horizontal van-step--waiting">
  <div class="van-step__title">
    B
  </div>
  <div class="van-step__circle-container">
    <i class="van-badge__wrapper van-icon van-icon-foo van-step__icon">
    </i>
  </div>
  <div class="van-step__line">
  </div>
</div>
`;

exports[`should render finish icon slot correctly 1`] = `
<div class="van-hairline van-step van-step--horizontal van-step--finish">
  <div class="van-step__title">
    A
  </div>
  <div class="van-step__circle-container">
    Custom Finish Icon
  </div>
  <div class="van-step__line">
  </div>
</div>
`;

exports[`should render icon slot correctly 1`] = `
<div class="van-steps van-steps--horizontal">
  <div class="van-steps__items">
    <div class="van-hairline van-step van-step--horizontal van-step--process">
      <div class="van-step__title van-step__title--active">
        B
      </div>
      <div class="van-step__circle-container">
        Custom Active Icon
      </div>
      <div class="van-step__line">
      </div>
    </div>
    <div class="van-hairline van-step van-step--horizontal van-step--waiting">
      <div class="van-step__title">
        A
      </div>
      <div class="van-step__circle-container">
        Custom Inactive Icon
      </div>
      <div class="van-step__line">
      </div>
    </div>
  </div>
</div>
`;

exports[`should render icon-prefix correctly 1`] = `
<div class="van-hairline van-step van-step--horizontal van-step--process">
  <div class="van-step__title van-step__title--active">
    B
  </div>
  <div class="van-step__circle-container">
    <i class="van-badge__wrapper custom-icon custom-icon-checked van-step__icon van-step__icon--active">
    </i>
  </div>
  <div class="van-step__line">
  </div>
</div>
`;
